parser/jsonparser/jsonparser.go

```diff
@@ -0,0 +1,30 @@
+package jsonparser
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// JsonParser parse a JSON string.
+type JsonParser struct{}
+
+// NewJsonParser returns a new json parser.
+func NewJsonParser() *JsonParser {
+	return &JsonParser{}
+}
+
+// ParseString implements the Parser interface.
+// The value in the map is not necessarily a string, so it needs to be converted.
+func (j *JsonParser) ParseString(line string) (map[string]string, error) {
+	var parsed map[string]interface{}
+	err := json.Unmarshal([]byte(line), &parsed)
+	if err != nil {
+		return nil, fmt.Errorf("json log parsing err: %w", err)
+	}
+
+	fields := make(map[string]string, len(parsed))
+	for k, v := range parsed {
+		fields[k] = fmt.Sprintf("%v", v)
```

Comment:
Would it possibly increase performance to add a type guard here (to only convert values that are _not already_ of type `string`)? Something like...

```suggestion
		if s, ok := v.(string); ok {
			fields[k] = s
		} else {
			fields[k] = fmt.Sprintf("%v", v)
		}
```

Reply:
Great! I made a benchstat. It's faster using type guard and the difference is statistically significant(p<0.05). You are right!👍 

```go
var parsed = map[string]interface{}{}

func initParsed() {
	for i := 0; i < 30000; i++ {
		parsed["time_local"+strconv.Itoa(i)] = "2021-02-03T11:22:33+08:00"
		parsed["request_length"+strconv.Itoa(i)] = 123
		parsed["request_method"+strconv.Itoa(i)] = "GET"
		parsed["request"+strconv.Itoa(i)] = "GET /order/2145 HTTP/1.1"
		parsed["body_bytes_sent"+strconv.Itoa(i)] = 518
		parsed["status"+strconv.Itoa(i)] = 200
		parsed["request_time"+strconv.Itoa(i)] = 0.544
		parsed["upstream_response_time"+strconv.Itoa(i)] = "0.543"
	}
}

func BenchmarkTypeGuard(b *testing.B) {
	initParsed()
	fields := make(map[string]string, len(parsed))
	b.StartTimer()
	for k, v := range parsed {
		if s, ok := v.(string); ok {
			fields[k] = s
		} else {
			fields[k] = fmt.Sprintf("%v", v)
		}
	}
}

// func BenchmarkTypeGuard(b *testing.B) {
// 	initParsed()
// 	fields := make(map[string]string, len(parsed))
// 	b.StartTimer()
// 	for k, v := range parsed {
// 		fields[k] = fmt.Sprintf("%v", v)
// 	}
// }
```
benchmark, modify function content between two tests.
```
go test -bench=BenchmarkTypeGuard -run=none -count=20 | tee old.txt
go test -bench=BenchmarkTypeGuard -run=none -count=20 | tee new.txt
```
benchstat result
```
$ benchstat old.txt new.txt
name         old time/op  new time/op  delta
TypeGuard-8  0.13ns ± 2%  0.12ns ± 6%  -10.28%  (p=0.000 n=19+19)
```